.buildkite/pipeline.py

```diff
@@ -0,0 +1,431 @@
+        data = {
+            "label": f"{self.project} {self.name}",
+            "command": self.step_command(),
+            "notify": [
+                {
+                    "github_commit_status": {
+                        "context": f"{self.project.title()}: {self.name}",
+                    }
+                }
+            ],
+            "agents": self.agent.create_entity(),
+            "artifact_paths": [
+                f"{self.project}/build/*.xml",
+                f"{self.project}/build/*.json",
+            ],
+        }
+        if self.env:
+            data["env"] = self.env
+        return data
+
+
+class Group:
+    """Buildkite Group object"""
+
+    def __init__(self, project: str, category: str, steps: list[Step]):
+        self.project: str = project
+        self.category: str = category
+        self.steps: list[Step] = steps
+
+    def __lt__(self, other):
+        return self.project < other.project
+
+    def create_entity(self) -> dict[str, Any]:
+        if len(self.steps) == 0:
+            return {}
+
+        data = {
+            "group": f"{self.project} {self.category}",
+            "key": f"{self.project}-{self.category}",
+            "steps": [step.create_entity() for step in self.steps],
+        }
+
+        return data
+
+
+class GitHelper:
+    def __init__(self):
+        self.files: list[str] = []
+
+    def get_pr_changeset(self) -> list[str]:
+        base_branch = os.getenv("BUILDKITE_PULL_REQUEST_BASE_BRANCH", "main")
+        diff_command = ["git", "diff", "--name-only", "{}...HEAD".format(base_branch)]
+        result = subprocess.run(diff_command, stdout=subprocess.PIPE)
+        if result.returncode == 0:
+            self.files = result.stdout.decode().splitlines()
+        else:
+            print(f"Detecting changed files failed, exiting [{result.returncode}]")
+            exit(result.returncode)
+        return self.files
+
+
+class BuildkitePipeline:
+    """Buildkite Pipeline object"""
+
+    def __init__(self, groups: list[Group] = None):
+        if groups is None:
+            groups = []
+        self.groups: list[Group] = groups
+
+    def create_entity(self):
+        data = {"steps": [group.create_entity() for group in self.groups]}
+        return data
+
+
+def is_pr() -> bool:
+    return os.getenv("BUILDKITE_PULL_REQUEST") != "false"
+
+
+def group_comment(group: Group) -> bool:
+    comment = os.getenv("GITHUB_PR_TRIGGER_COMMENT")
+    if comment:
+        # the comment should be a subset of the values
+        # in .buildkite/pull-requests.json
+        # TODO: change /test
+        comment_prefix = "buildkite test"
+        if group.category == "mandatory":
+            # i.e: /test filebeat
+            return comment_prefix + " " + group.project in comment
+        else:
+            # i.e: test filebeat extended
+            return (
+                comment_prefix + " " + group.project + " " + group.category in comment
+            )
+
```

Comment:
Otherwise

`return False`? Or what's the expected behaviour if no `else` condition?

Reply:
the triggering with comments and running on branches is still WIP. We moved forward with integrating this and iterated over improvement on the specific parts.